packages/oauth/oauth-provider/src/client/client-auth-check.ts

```diff
@@ -0,0 +1,53 @@
+import { CLIENT_ASSERTION_TYPE_JWT_BEARER } from '@atproto/oauth-types'
+import { InvalidGrantError } from '../errors/invalid-grant-error.js'
+import { ClientAuth } from './client-auth.js'
+import { ClientId } from './client-id.js'
+import { Client } from './client.js'
+
+/**
+ * Ensures that a {@link ClientAuth} used by a current request matches a
+ * {@link previous} authentication method currently.
+ */
+export function clientAuthCheck<C extends ClientAuth>(
+  client: Client,
+  clientAuth: C,
+  previous: {
+    clientId: ClientId
+    clientAuth: ClientAuth
+  },
+): asserts previous is { clientId: ClientId; clientAuth: C } {
+  if (previous.clientId !== client.id) {
+    throw new InvalidGrantError(`Token was not issued to this client`)
+  }
+
+  switch (previous.clientAuth.method) {
+    case CLIENT_ASSERTION_TYPE_JWT_BEARER:
+      if (clientAuth.method !== previous.clientAuth.method) {
+        throw new InvalidGrantError(
+          `Client authentication method mismatch (expected ${previous.clientAuth.method})`,
+        )
+      }
+      if (
+        clientAuth.kid !== previous.clientAuth.kid ||
+        clientAuth.alg !== previous.clientAuth.alg ||
+        clientAuth.jkt !== previous.clientAuth.jkt
```

Comment:
Just confirming: the client key's `jkt` is calculated by oauth-provider and doesn't come from the JWK itself, right?  I know this as a DPoP mechanism so just making sure I understand its usage here.

Reply:
That's right. Computed here:

https://github.com/bluesky-social/atproto/blob/a1feaf84963afbaa323aaa79138551d1e1e8841b/packages/oauth/oauth-provider/src/client/client.ts#L248